File: src/test/java/lavahack/kevin/benchmark/benchmarks/KevinBenchmark.java
package lavahack.kevin.benchmark.benchmarks;

import lavahack.kevin.EventBus;
import lavahack.kevin.Listener;
import lavahack.kevin.benchmark.BenchmarkEvent;
import lavahack.kevin.benchmark.Constants;
import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.Setup;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.infra.Blackhole;

@State(Scope.Benchmark)
public class KevinBenchmark {
    private EventBus<BenchmarkEvent> bus;

    @Setup
    public void setup() {
        bus = new EventBus<>();

        for (int i = 0; i < Constants.LISTENERS; i++) {
            bus.subscribe(new Listener<>(event -> {
                event.blackhole().consume(Integer.bitCount(Integer.parseInt("123")));
            }));
        }
    }

    @Benchmark
    public void benchmark(Blackhole blackhole) {
        bus.post(new BenchmarkEvent(blackhole));
    }
}

File: benchmarks/src/jmh/java/benchmarks/KevinBenchmark.kt
package benchmarks

import lavahack.kevin.Listener
import org.openjdk.jmh.annotations.*
import org.openjdk.jmh.infra.Blackhole
import java.util.*
import java.util.concurrent.TimeUnit

@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@State(Scope.Benchmark)
@Fork(value = 1)
@Warmup(iterations = 5, time = 5)
@Measurement(iterations = 5, time = 5)
open class KevinBenchmark {
    companion object {
        @JvmStatic
        fun prepare() {
            for(i in 0..<Constants.LISTENERS) {
                val listener = Listener<KevinEvent>(Random().nextInt()) {
                    it.blackhole.consume(Integer.bitCount(Integer.getInteger("123")))
                }

                KevinEvent.BUS.subscribe(listener)
            }
        }

    }

    @Benchmark
    fun benchmark(
        blackhole : Blackhole
    ) {
        val event = KevinEvent(blackhole)

        KevinEvent.BUS.post(event)
    }
}